### PR TITLE
ci: fix CI after bazel 8 release

### LINKED ISSUE #2378: Support Bazel 8
I was trying to run our test with Bazel 8 and have been encountering a variety of problems.

The overall impression I'm getting is that the the core code is working, but various dev dependencies aren't.

Debugging this has been rather tough. Usually this is just a matter of updating some dependencies, but I've yet to find a combination of dependencies that works with both Bazel 7 and Bazel 8.

Issues I've seen thus far:

* [x] WORKSPACE.bzlmod is non-empty. It needs to become empty as WORKSPACE files become no-ops
* [ ] gazelle has `native.sh_binary` references: https://github.com/bazel-contrib/bazel-gazelle/issues/1959
* [x] We need to update rules_bazel_integration_test version to fix some native references.
* [ ] rules_cc shenanigans -- it seems that between 0.0.9 and 0.0.14 there have been various changes to how to reference bzl files. This ends up breaking doc building and the bzl_library deps. What we'll probably want to do is stop using cc:defs.bzl and start using the more specific loads instead.
* [ ] Some issue with stardoc. Upgrading to 0.7.1 seems to fix it, _but_ the release notes for stardoc say 0.7 _requires_ Bazel 7. This will probably be OK; stardoc is a dev dependency, and only otherwise used via sphinxdocs, which is use-at-your-own-risk.

Various notes:
rules_cc centric:
* **rules_cc 0.0.9 can't be used with bazel 8: it lacks some fixes**
* **rules_cc 0.0.10 can't be used with bazel 6: it has a dep on stardoc 7, which requires (in its MODULE definition) bazel 7+**
* rules_cc 0.0.11, 0.0.12, and 0.0.13 can't be used with Bazel 6: it tries to use `extension_metadata(reproducible=True)`
* rules_cc 0.0.14 might be usable with Bazel 6 (mod graph passes) and relnotes say they restored bazel 6 compatibility
* rules_cc 0.0.14 with protobuf 24.4 doesn't work -- protobuf loads `@rules_cc//:defs.bzl#cc_proto_library`, which was removed in rules_cc 0.0.14`. This was fixed in protobuf 27.0.
* So basically, we must use rules_cc 0.0.14 and protobuf 27.0 or higher together.

stardoc centric:
* **Bazel 6 with bzlmod and stardoc 0.7 is a problem: Bazel will fail during module phase because stardoc requires (in its MODULE) Bazel 7.**
* The stardoc 0.7 requiring bazel 7 comes from rules_jvm_external: rules_jvm_external has a `use_repo_rule` call, which is Bazel 7 only.

rules_java centric:
* rules_java before 8.2.0 and using the remote jdsk can't be used with rules_cc 0.0.14 because of the rules_cc cc_proto_library removal.
* **Usage of the remotejdk only appears to be done by our examples with the coverage command.**
* rules_java 8.2.0+ fixes the rules_cc issue, but can't be used because of a bug defining an android rule which fails due to some private API being used.
* rules_java 8.2.0 requires Bazel 7.3.2 or higher (MODULE checks bazel version)
* rules_java 8.3.0 works, but:
* **rules_java 8.3.0 doesn't support Bazel 6** (MODULE checks bazel version)

proto centric
* protobuf 24 can't be used because of the rules_cc 0.0.14 removal of cc_proto_library
* protobuf 28 is _sort of_ the minimum, but our CI is unhappy with that version because protobuf set `-Werror` (treat warnings as errors), and the debian jobs then fail due to various warnings when compiling protobuf. Other OSes are OK
* protobuf 29 fixes the `-Werror` problem, _but_ that version isn't fully released. So 29.0-rc1 is the next available version.
* Alternatively, we could use `toolchains_protoc` in our dev builds, which should avoid having to compile protoc entirely.

### PR DESCRIPTION
**Pull request by aignas**
Summary:
* Remove the old `7.x` config from all `.bazelrc`.
* Drop bazel 6 from example testing/support. The `.bazelrc`
  for enabling `WORKSPACE` cannot work across bazel 6,7,8.
* Add missing `BUILD.bazel` file for integration tests.
* Remove an integration test runner for `bazel 6.x`.
* RBE test for bazel 8 is still not working. 
* bump `rules_java` for internal WORKSPACE dependencies to
  a version that supports `8.x`.
* start running bazel-in-bazel integration tests using bazel `8.x`.
* until bazel-contrib/rules_bazel_integration_test#414 is merged,
  we need to use bazel 7 for the delete_packagese pre-commit hook,
  not sure about how to track this as the `cgrindel/bazel-lib` needs
  a new version.

Fixes #2378

### CODE REVIEW

**Comment by aignas** [comment]
Didn't have time to finish this. Tried doing:
* Bumping minimum tested version of bazel to 7.4.1
* Remove the `7.x` config from bazelrc
* Add `--enable_workspace` to workspace only examples (actually, should we move them to internal or somewhere, so that they don't advertise WORKSPACE support anymore?).

The build is failing with `rules_java` and it seems that it might be defined somewhere early in the process for the WORKSPACE builds.

**Comment by aignas** [comment]
Reverted the changes and trying a more minimal stop-gap solution.

**Comment by rickeylev** [comment]
> The build is failing with rules_java ...
> ERROR: Error computing the main repository mapping: cannot load '@@rules_java//java:rules_java_deps.bzl': no such file

Ugh, that error. I've seen that a ton while trying to get Bazel 9 workspace working. From what I've been able to tell, it means rules_java needs to be upgraded.

Ah, I see we have CI coverage for bazel 8 bzlmod, but not workspace :facepalm: .

**Comment by rickeylev** [comment]
This is what I have in some WIP stuff for bazel 9:

```
    http_archive(
        name = "rules_java",
        urls = [
            "https://github.com/bazelbuild/rules_java/releases/download/8.6.2/rules_java-8.6.2.tar.gz",
        ],
        sha256 = "a64ab04616e76a448c2c2d8165d836f0d2fb0906200d0b7c7376f46dd62e59cc",
    )
```

There may also be a change to the `rules_java_dependencies()` call in internal_dev_setup.bzl needed --  `load("@rules_java//java:deps.bzl", "rules_java_dependencies")` might be needed. (not entirely sure)

File: WORKSPACE [hunk 1]
```diff
@@ -21,6 +21,21 @@ load("//:internal_dev_deps.bzl", "rules_python_internal_deps")
 
 rules_python_internal_deps()
 
+load("@rules_java//java:rules_java_deps.bzl", "rules_java_dependencies")
```
Comment (aignas) [inline WORKSPACE:24]: NOTE: I could not get this setup to work from `internal_dev_setup.bzl`.

**Review by rickeylev — APPROVED** [review]

**Review by rickeylev — APPROVED** [review]

**Pull request merged**